alibi/explainers/similarity/base.py

```diff
@@ -149,18 +156,19 @@ def _compute_adhoc_similarity(self, grad_X: np.ndarray) -> np.ndarray:
         grad_X
             Gradients of the test instances.
         """
-        scores = np.zeros((grad_X.shape[0], self.X_train.shape[0]))
+        scores = np.zeros((len(grad_X), len(self.X_train)))
+        X: Union[np.ndarray, List[Any]]
         for i, (X, Y) in tqdm(enumerate(zip(self.X_train, self.Y_train)), disable=not self.verbose):
-            grad_X_train = self._compute_grad(X[None], Y[None])
+            X = X[None] if hasattr(self.X_train, 'shape') else [X]  # type: ignore[call-overload]
```

Comment:
Is there a reason you check if `X_train` has a shape attribute here instead of checking its a `np.ndarray`?

Reply:
This is just a very quick prototype and I choose to go checking the `shape` attribute for consistency. There is another place in the code in which I need to check if it's either `np.ndarray` or `tf.Tensor` or `pt.Tensor`. I think the problem with checking whether is either `tf.Tensor` or `pt.Tensor` is whether or not the dependencies are installed. So I think I need to check that first.